nodes can't stop allocations if they've changed node pools

Cross-posted from https://github.com/hashicorp/nomad/issues/27961#issuecomment-4631069472

@optiz0r wrote:

> We moved all nodes from default node pool into separate node pools a few months ago, along with changing namespace default node_pool, and the default node pool became empty. Some jobs (inheriting the namespace default node pool) were never properly restarted after this change, so the running allocations were still associated with the default node pool while still running on nodes which were moved into a different node pool.
> 
> Doing an upgrade from 1.10.8(+ent) to 2.0.2(+ent) today got nomad into a state where those jobs failed to reschedule due to "no eligible nodes" (in default node pool). Stopping and starting the jobs reassociated them with the namespace-default node pool which did contain resources. But the allocations stuck in pending state with the new node logging errors being unable to fetch the previous allocation ID.

```
Jun 05 11:40:16 worker2 nomad[209858]:     2026-06-05T11:40:16.691+0100 [ERROR] client.rpc: error performing RPC to server: error="rpc error: rpc error: Unknown allocation \"4dbffbe7-fca9-b12c-c09e-3f82eb5c22b8\"" rpc=Alloc.GetAlloc server=10.x.x.x:4647
Jun 05 11:40:16 worker2 nomad[209858]:     2026-06-05T11:40:16.691+0100 [ERROR] client.rpc: error performing RPC to server which is not safe to automatically retry: error="rpc error: rpc error: Unknown allocation \"4dbffbe7-fca9-b12c-c09e-3f82eb5c22b8\"" rpc=Alloc.GetAlloc server=10.x.x.x:4647
```

> The node on which the jobs had been running was logging the same errors in this thread at high rates:

```Jun 05 12:02:38 worker1 nomad[909]:     2026-06-05T12:02:38.575+0100 [ERROR] client: error querying updated allocations: error="rpc error: Permission denied"
Jun 05 12:02:38 worker1 nomad[909]:     2026-06-05T12:02:38.583+0100 [ERROR] client.rpc: error performing RPC to server: error="rpc error: Permission denied" rpc=Alloc.GetAllocs server=10.x.x.x:4647
```

> It was not possible to drain the node, and even after rebooting the node it was still logging permission errors. To clear the fault, I stopped nomad agent, nuked the datadir (containing client-id and state.db) and rejoined it to the cluster as a new node. After some time, nomad correctly recovered all jobs, and permission errors are no longer being logged. 

This looks like two different issues:
* When the client comes back online with a new node pool, it makes `Alloc.GetAllocs` calls to get updates on allocations it's supposed to have, but now those allocations are assigned to the wrong pool because they belong to an older version of the job. So the node is asking for the alloc, the server says "that's not in your node pool" and disallows it. This is a fairly significant bug.
* The `Alloc.GetAlloc` (singular) RPC call is from the alloc watcher, which is how the node migrates workloads from one node to another. This _should_ fail, making it impossible to do the best-effort migration between node pools. (This was a known tradeoff.) But it shouldn't be looping there... it should fail once and then stop watching.

--- a/.changelog/28110.txt
+++ b/.changelog/28110.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+auth: Fixed a bug where nodes could not sync allocations placed on them after being moved to a different node pool
+```

--- a/.semgrep/rpc_endpoint.yml
+++ b/.semgrep/rpc_endpoint.yml
@@ -157,7 +157,7 @@ rules:
             return structs.ErrPermissionDenied
           }
           ...
-          if err := $A.$B.AuthorizeClientAllocation(aclObj, ..., ...); err != nil {
+          if err := $A.$B.AuthorizeClientAllocation(..., aclObj, ..., ...); err != nil {
             return err
           }
           ...

--- a/nomad/acl.go
+++ b/nomad/acl.go
@@ -29,11 +29,12 @@ func (s *Server) ResolveAuthorizedClientNodePoolByNodeID(aclObj *acl.ACL, nodeID
 }
 
 func (s *Server) AuthorizeClientAllocation(
+	identity *structs.AuthenticatedIdentity,
 	aclObj *acl.ACL,
 	alloc *structs.Allocation,
 	allowNsOp func(*acl.ACL, string) bool,
 ) error {
-	return s.auth.AuthorizeClientAllocation(aclObj, alloc, allowNsOp)
+	return s.auth.AuthorizeClientAllocation(identity, aclObj, alloc, allowNsOp)
 }
 
 func (s *Server) ResolveAuthorizedClientNodePoolByServiceRegistrationID(

--- a/nomad/alloc_endpoint.go
+++ b/nomad/alloc_endpoint.go
@@ -165,7 +165,7 @@ func (a *Alloc) GetAlloc(args *structs.AllocSpecificRequest,
 				reply.Alloc = out
 
 				// Re-check namespace in case it differs from request.
-				if err := a.srv.AuthorizeClientAllocation(aclObj, out, allowNsOp); err != nil {
+				if err := a.srv.AuthorizeClientAllocation(args.GetIdentity(), aclObj, out, allowNsOp); err != nil {
 					return structs.NewErrUnknownAllocation(args.AllocID)
 				}
 
@@ -231,7 +231,7 @@ func (a *Alloc) GetAllocs(args *structs.AllocsGetRequest,
 					continue
 				}
 
-				if err := a.srv.AuthorizeClientAllocation(aclObj, out, nil); err != nil {
+				if err := a.srv.AuthorizeClientAllocation(args.GetIdentity(), aclObj, out, nil); err != nil {
 					return err
 				}
 
@@ -492,7 +492,7 @@ func (a *Alloc) SignIdentities(args *structs.AllocIdentitiesRequest, reply *stru
 					continue
 				}
 
-				if err := a.srv.AuthorizeClientAllocation(aclObj, out, nil); err != nil {
+				if err := a.srv.AuthorizeClientAllocation(args.GetIdentity(), aclObj, out, nil); err != nil {
 					return err
 				}
 

--- a/nomad/alloc_endpoint.go
+++ b/nomad/alloc_endpoint.go
@@ -164,7 +164,9 @@ func (a *Alloc) GetAlloc(args *structs.AllocSpecificRequest,
 				}
 				reply.Alloc = out
 
-				// Re-check namespace in case it differs from request.
+				// Authorize the caller against the alloc we found: allow the
+				// node the alloc is placed on, client ACLs for the alloc's
+				// node pool, or namespace read-job as a fallback.
 				if err := a.srv.AuthorizeClientAllocation(args.GetIdentity(), aclObj, out, allowNsOp); err != nil {
 					return structs.NewErrUnknownAllocation(args.AllocID)
 				}

--- a/nomad/auth/auth.go
+++ b/nomad/auth/auth.go
@@ -665,12 +665,17 @@ func (s *Authenticator) ResolveAuthorizedClientNodePoolByNodeID(aclObj *acl.ACL,
 // fall back to namespace-based authorization when client-scoped authorization
 // does not apply.
 func (s *Authenticator) AuthorizeClientAllocation(
+	identity *structs.AuthenticatedIdentity,
 	aclObj *acl.ACL,
 	alloc *structs.Allocation,
 	allowNsOp func(*acl.ACL, string) bool,
 ) error {
 	if alloc == nil || alloc.Job == nil {
 		return structs.ErrPermissionDenied
+	}
+
+	if identity != nil && AuthorizeSameNode(identity, alloc.NodeID) == nil {
+		return nil
 	}
 
 	if aclObj.AllowClientOp(alloc.Job.NodePool) {

--- a/nomad/auth/auth.go
+++ b/nomad/auth/auth.go
@@ -660,10 +660,10 @@ func (s *Authenticator) ResolveAuthorizedClientNodePoolByNodeID(aclObj *acl.ACL,
 	return resolveAuthorizedClientNodePoolByNodeID(snap, aclObj, nodeID)
 }
 
-// AuthorizeClientAllocation returns ErrPermissionDenied unless aclObj is
-// authorized for alloc's node pool. If allowNsOp is provided, callers may
-// fall back to namespace-based authorization when client-scoped authorization
-// does not apply.
+// AuthorizeClientAllocation returns ErrPermissionDenied unless the identity
+// is the node the alloc is placed on or aclObj is authorized for alloc's
+// node pool. If allowNsOp is provided, callers may fall back to
+// namespace-based authorization when client-scoped authorization does not apply.
 func (s *Authenticator) AuthorizeClientAllocation(
 	identity *structs.AuthenticatedIdentity,
 	aclObj *acl.ACL,
@@ -674,7 +674,7 @@ func (s *Authenticator) AuthorizeClientAllocation(
 		return structs.ErrPermissionDenied
 	}
 
-	if identity != nil && AuthorizeSameNode(identity, alloc.NodeID) == nil {
+	if alloc.NodeID != "" && AuthorizeSameNode(identity, alloc.NodeID) == nil {
 		return nil
 	}
 

--- a/nomad/auth/auth_test.go
+++ b/nomad/auth/auth_test.go
@@ -1744,6 +1744,35 @@ func TestResolveAuthorizedClientNodePoolHelpers(t *testing.T) {
 			alloc,
 			nil,
 		), structs.ErrPermissionDenied)
+
+		// A node authenticating with node identity claims rather than a
+		// client secret must also be authorized for its own allocs.
+		claimsIdentity := &structs.AuthenticatedIdentity{
+			Claims: &structs.IdentityClaims{
+				NodeIdentityClaims: &structs.NodeIdentityClaims{
+					NodeID:   node.ID,
+					NodePool: node.NodePool,
+				},
+			},
+		}
+		must.NoError(t, auth.AuthorizeClientAllocation(
+			claimsIdentity,
+			acl.NewClientACL("other-pool"),
+			alloc,
+			nil,
+		))
+
+		// An unplaced alloc must not match a non-node identity, whose
+		// authenticated node ID is also empty.
+		unplacedAlloc := mock.Alloc()
+		unplacedAlloc.Job.NodePool = "other-pool"
+		unplacedAlloc.NodeID = ""
+		must.ErrorIs(t, auth.AuthorizeClientAllocation(
+			&structs.AuthenticatedIdentity{},
+			aclObj,
+			unplacedAlloc,
+			nil,
+		), structs.ErrPermissionDenied)
 	})
 
 	t.Run("resolve by service registration id", func(t *testing.T) {

--- a/nomad/auth/auth_test.go
+++ b/nomad/auth/auth_test.go
@@ -1713,14 +1713,37 @@ func TestResolveAuthorizedClientNodePoolHelpers(t *testing.T) {
 	t.Run("authorize client allocation", func(t *testing.T) {
 		alloc := mock.Alloc()
 		alloc.Job.NodePool = node.NodePool
+		alloc.NodeID = node.ID
 
-		must.NoError(t, auth.AuthorizeClientAllocation(aclObj, alloc, nil))
-		must.ErrorIs(t, auth.AuthorizeClientAllocation(acl.NewClientACL("other-pool"), alloc, nil), structs.ErrPermissionDenied)
+		must.NoError(t, auth.AuthorizeClientAllocation(nil, aclObj, alloc, nil))
+		must.ErrorIs(t, auth.AuthorizeClientAllocation(nil, acl.NewClientACL("other-pool"), alloc, nil), structs.ErrPermissionDenied)
 		must.NoError(t, auth.AuthorizeClientAllocation(
+			nil,
 			acl.NewClientACL("other-pool"),
 			alloc,
 			func(_ *acl.ACL, ns string) bool { return ns == alloc.Namespace },
 		))
+
+		// Test identity-based authorization when node pool mismatches
+		matchingIdentity := &structs.AuthenticatedIdentity{
+			ClientID: node.ID,
+		}
+		must.NoError(t, auth.AuthorizeClientAllocation(
+			matchingIdentity,
+			acl.NewClientACL("other-pool"),
+			alloc,
+			nil,
+		))
+
+		mismatchIdentity := &structs.AuthenticatedIdentity{
+			ClientID: "other-node",
+		}
+		must.ErrorIs(t, auth.AuthorizeClientAllocation(
+			mismatchIdentity,
+			acl.NewClientACL("other-pool"),
+			alloc,
+			nil,
+		), structs.ErrPermissionDenied)
 	})
 
 	t.Run("resolve by service registration id", func(t *testing.T) {

